module: fix inconsistency between load and _findPath

lundibundi opened the pull request:
Files with multiple extensions are not handled by require-module system
therefore if you have file 'file.foo.bar' and require('./file') it won't
be found even while using require.extensions['foo.bar'] but before this
commit if you have require.extensions['foo.bar'] and
require.extensions['bar'] set then the latter will be called if you do
require('./file') but if you remove the former ('foo.bar')
property it will fail.
This commit makes it always fail in such cases.

Fixes: https://github.com/nodejs/node/issues/4778

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -216,6 +216,11 @@ function tryExtensions(p, exts, isMain) {
 
+function readExtensions() {
+  return Object.keys(Module._extensions)
+    .map((ext) => path.extname(ext) || ext);
```
Can we avoid the functional array methods and use a regular loop instead?

lundibundi replied:
I'm not against this but imo this looks much better than regular loop would. As for the performance implications this is not a hot code and also in the newest v8 AFAIK functional operations have the same or at least similar performance to regular loops.
Though if you want I'll change it. I think it will be as follows.
```js
const exts = Object.keys(Module._extensions);
const result = new Array(exts.length);
for (let i = 0; i < exts.length; ++i)
  result[i] = path.extname(exts[i]) || exts[i];
return result;
```

mscdex replied:
I think you could simplify if further and make it even more efficient than a straight conversion with:

```js
const exts = Object.keys(Module._extensions);
for (let i = 0; i < exts.length; ++i)
  exts[i] = path.extname(exts[i]) || exts[i];
return exts;
```

or avoid reassigning the same value:

```js
const exts = Object.keys(Module._extensions);
for (let i = 0; i < exts.length; ++i) {
  const extName = path.extname(exts[i]);
  if (extName)
    exts[i] = extName;
}
return exts;
```

lundibundi replied:
Oh, yeah. Functional thinking got too far =).

jdalton commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -216,6 +216,13 @@ function tryExtensions(p, exts, isMain) {
+  for (var i = 0; i < exts.length; ++i)
+    exts[i] = path.extname(exts[i]) || exts[i];
+  return exts;
+}
```
In the example case of `require.extensions` of `.foo.bar` and `.bar` this will return an array of `['.bar', '.bar]`?
So if you `delete require.extensions[".bar"]` files like `test-extensions.foo.bar` will still be found but their compiler at `require.extensions[".foo.bar"]` won't be used for the module since `_compile` will use `extname(filename)` to and get `.bar` which doesn't exist on `require.extensions` and so will default to `.js`.

jdalton commented:
This patch makes additions like `require.extensions[".foo.bar"]` findable but not runnable as it defaults to `require.extensions[".js"]` since the `extname(filename)` is `".bar"` and does not have an entry in `require.extensions`. This means this PR isn't really finished. Finding a module is great but it needs to use the correct compiler too.

lundibundi commented:
@jdalton PTAL https://github.com/nodejs/node/pull/22382#discussion_r211886584, I've changed the implementation to ignore multiple-extensions-type (as this is from what I understand compiler does via https://github.com/nodejs/node/blob/master/lib/internal/modules/cjs/loader.js#L597). The idea of this PR is to forbid what was previously accepted but is incorrect (if you have `require.extensions[".foo.bar"] = 'anything' and require.extensions[".bar"] = () => {}` then the latter was called upon `require('./file')` assuming we had `file.foo.bar` file). It doesn't make them findable, it does the contrary and makes the not findable.

I've also added the test for the use-case you've described (have `'.foo.bar'` in `extensions` and call `require('./test-extensions.foo')` which should fail).

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16797/
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/229/

lundibundi commented:
Resume: https://ci.nodejs.org/job/node-test-pull-request/16842/

addaleax commented:
@lundibundi Do you want to land this? It seems ready to me. :)

lundibundi commented:
@addaleax thanks for reminding I kinda forgot =)
Wanted to make CI happy finally.
Another try: https://ci.nodejs.org/job/node-test-pull-request/16947/

addaleax commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/16959/

lundibundi commented:
Another resume: https://ci.nodejs.org/job/node-test-pull-request/16973/
tough fight :upside_down_face:.

addaleax commented:
@lundibundi I’m not sure I understand – all of the last three CIs (16959, 16973, 16979) were good…?

lundibundi commented:
Oh, I kind of didn't check that the failed test was with node help options. :upside_down_face: 

Though I wanted to land this already, but `node-core-utils` (`git-node`) for some reason gets stuck on network operations and cannot reset my master with upstream/master...
Will try again, just skip that step I guess, I can always manually rebase on upstream/master.

addaleax commented:
Landed in 1b92214d097358040efb7d3ec5dff1736f364bc0
